### PR TITLE
[Merged by Bors] - chore(MeasureTheory/Group/Arithmetic): tag more lemmas with `fun_prop`

### PR DESCRIPTION
**Pull request by YaelDillies**
... and rewrite `Measurable.mul'` and `Measurable.div'` to be in compositional form

From PFR

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by grunweg — COMMENTED** [review]
Thanks for this PR; tagging more lemmas with fun_prop always seems good.
I wonder about the changes to `Group/Arithmetic.lean`: there's an existing lemma `Measurable.mul` tagged fun_prop; and you're tagging the primed version as well. Both look equivalent, up to a beta reduction.
@lecopivo Is there a pitfall to be aware of, or are these fine to tag?

If that question has a positive answer, feel free to maintainer merge on my behalf.

**Comment by lecopivo** [comment]
tldr; Currently `fun_prop` does not simplify `(f + g) x` to `f x + g x` as a consequence `Measurable.mul` and `Measurable.mul'` are necessary but `Measurable.mul'` should be stated in more general form.

The theorem
```
theorem Measurable.mul [MeasurableMul₂ M] (hf : Measurable f) (hg : Measurable g) :
    Measurable fun a => f a * g a
```
talks about measurability of `HMul.hMul` in its first and second argument. This theorem is stated in so called *compositional form*. See [doc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Tactic/FunProp.html) which talks about *compositional* vs *uncurried* form.

The theorem 
```
theorem Measurable.mul' [MeasurableMul₂ M] (hf : Measurable f) (hg : Measurable g) :
    Measurable (f * g)
```
talks about measurability of  `HMul.hMul` in its third argument. This theorem is stated in so called *uncurried* form.

In `Measurable.mul'`, multiplication is considered as function of three arguments and the theorem talks only about measurability in the third argument. Ideally we would have a theorem talking about measurability in all three arguments. Because multiplication, as a function of three arguments, is high-order function the only way `fun_prop` can properly work with it is if it has the theorem stated in the *compositional form*:
```
@[fun_prop]
theorem Measurable.mul''
    {W M α : Type*} [MeasurableSpace W] [MeasurableSpace M] [Mul M] {m : MeasurableSpace α} [MeasurableMul₂ M]
    {f g : W → α → M} {h : W → α}
    (hf : Measurable (↿f)) (hg : Measurable (↿g)) (hh : Measurable h):
    Measurable (fun w => ((f w) * (g w)) (h w)) := by simp; fun_prop
```

If you want to consider multiplication as a function of four arguments you should add additional theorem:
```
@[fun_prop]
theorem Measurable.mul'''
    {W M α β : Type*} [MeasurableSpace W] [MeasurableSpace M] [Mul M] {m : MeasurableSpace α} {m' : MeasurableSpace β} [MeasurableMul₂ M]
    {f g : W → β → α → M} {h : W → α} {h' : W → β}
    (hf : Measurable (↿f)) (hg : Measurable (↿g)) (hh : Measurable h) (hh' : Measurable h'):
    Measurable (fun w => ((f w) * (g w)) (h' w) (h w)) := by simp; fun_prop
```
and so on.

Just a reminder, anytime you are unsure if it is a good idea to add theorem or not then turn on `set_option trace.Meta.Tactic.fun_prop.attr true` before declaring `fun_prop` theorem and look at info.

**Comment by YaelDillies** [comment]
Thanks for the info. I've rewritten `Measurable.mul'` and `Measurable.div'` in compositional form. I have left their `AEMeasurable` version as future work as I then need to start messing around with measures.

File: Mathlib/MeasureTheory/Function/Intersectivity.lean [hunk 1]
```diff
@@ -61,8 +61,9 @@ lemma bergelson' {s : ℕ → Set α} (hs : ∀ n, MeasurableSet (s n)) (hr₀ :
   have hfapp : ∀ n a, f n a = (↑(n + 1))⁻¹ * ∑ k in Finset.range (n + 1), (s k).indicator 1 a := by
     simp only [f, Pi.natCast_def, Pi.smul_apply, Pi.inv_apply, Finset.sum_apply, eq_self_iff_true,
     forall_const, imp_true_iff, smul_eq_mul]
-  have hf n : Measurable (f n) := Measurable.mul' (@measurable_const ℝ≥0∞ _ _ _ (↑(n + 1))⁻¹)
-      (Finset.measurable_sum' _ fun i _ ↦ measurable_one.indicator <| hs i)
+  have should_be_found_by_fun_prop k :
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/Intersectivity.lean:64]: Yes, but actually I think I just forgot that `fun_prop` doesn't discharge with `assumption` by default. I will try again when I'm at a computer

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/Intersectivity.lean]: Solved!

**Comment by grunweg** [comment]
Thanks!
maintainer merge
